New method: groupBy (128 B)

BabkinAleksandr opened the pull request:
[Ramda groupBy](http://ramdajs.com/docs/#groupBy)

Tests was repeated for 100 times for each method. Lengths was: 100/10000/100000

`reduce`:
<img width="367" alt="groupby - reduce" src="https://user-images.githubusercontent.com/25172855/37326300-836138b2-26a2-11e8-9182-85f5fa3d86f0.png">

`for`:
<img width="362" alt="groupby - for" src="https://user-images.githubusercontent.com/25172855/37326320-97d472b4-26a2-11e8-9875-6494172b34f4.png">

`while`:
<img width="366" alt="groupby - while" src="https://user-images.githubusercontent.com/25172855/37326317-94a76376-26a2-11e8-9da0-56f194b7fb49.png">

Chosed `for`, because it's more important to be not slower, than faster, I think. But tests goes not very accurate. If we use array lengths 100/10000/100000, 100-length gives us +148%, if we use 10/100/1000, 100-length gives us -50%. For example, here you could see the same method, run 100 times with 10/100/1000:

`for`:
<img width="356" alt="aleksandrbabkin_aleksandrs-macbook-pro____documents_nanoutils" src="https://user-images.githubusercontent.com/25172855/37326878-1de8d992-26a5-11e8-99f7-63e80afef221.png">

Beraliv commented on lib/groupBy/groupBy.test.js:
```diff
@@ -0,0 +1,10 @@
+
+  const result = groupBy(func, list)
+
+  expect(result).toMatchObject({ even: [2, 4, 6, 8, 10], 'odd': [1, 3, 5, 7, 9] })
```
replace `'odd'` with just `odd`

BabkinAleksandr replied:
done

Beraliv commented on lib/groupBy/groupBy.performance.js:
```diff
@@ -0,0 +1,25 @@
+    if (i < 100) list1.push(i)
+    if (i < 10000) list2.push(i)
+    list3.push(i)
+  }
```
you should choose either `1e9, 1e8` and `1e7` and use `type: array_size` or change type and use `1e5, 1e4` and `1e3`

Beraliv commented on lib/groupBy/index.js:
```diff
@@ -0,0 +1,12 @@
+    var key = func(list[i])
+    obj.hasOwnProperty(key)
+      ? obj[key].push(list[i])
+      : obj[key] = [list[i]]
```
maybe replace with `obj[key] = (obj[key] || []).concat(list[i])`? Or is it slower?

BabkinAleksandr replied:
tested `concat`, it's slower, because it creates new array every time and redefining object[key] every time. Test crushes with 1e3/1e4/1e4 arrays

Beraliv replied:
maybe replace `obj.hasOwnProperty(key)` with `obj[key]`?

BabkinAleksandr replied:
`obj[key]` will search through prototype chain. I think we do not want such behavior, but It's opened question

BabkinAleksandr commented:
Fixed performance test, added new time test type.

Here is time test for `for`:
<img width="317" alt="aleksandrbabkin_aleksandrs-macbook-pro____documents_nanoutils" src="https://user-images.githubusercontent.com/25172855/37329073-5a8893da-26ad-11e8-863d-497ff01274b0.png">

Also, fixed `forEachObjIndexed` time test

BabkinAleksandr commented:
Reduced by 2 B
